### PR TITLE
HD Reflection System

### PR DESCRIPTION
**Pull request by fredpointzero**
### Purpose of this PR

Handle properly the baking and rendering of Reflection and Planar probes.
Provides an API to bake or render Reflection or Planar probes from C#.

---
### Release Notes

Added new Reflection System to control Reflection and Planar probe baking and rendering.

---
### Testing status
[Katana link](https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2FprobeRenderAPI&automation-tools_branch=add-platform-filter&unity_branch=trunk#)

**Manual Tests**:
 - Test continuous and manual baking on single/multiple scenes
 - Test baking of baked/custom reflection/planar probes
 - Test influence of probes

**Automated Tests**:
 - Added EditMode tests to migrate probes from previous data layouts to the last one.

---
### Overall Product Risks
**Technical Risk**: High

**Halo Effect**: Medium
All environment lighting are involved.

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 1]
```diff
@@ -119,6 +119,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [4.0.0-preview] - 2018-09-28
 
 ### Added
+- Added controls for linear fade at the boundary of density volumes
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: changelog are at the wrong place now :)

File: com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.Cs.meta [hunk 2]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.Cs.meta]: weird name "Cs" is it a mistake?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 3]
```diff
@@ -159,7 +159,7 @@ public enum StencilBitMask
             Clear                           = 0,    // 0x0
             LightingMask                    = 7,    // 0x7  - 3 bit
             Decals                          = 8,    // 0x8  - 1 bit
-            DecalsForwardOutputNormalBuffer = 16,   // 0x10 - 1 bit
+            DecalsForwardOutputNormalBuffer = 16,   // 0x10  - 1 bit
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: remove this space

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 4]
```diff
@@ -1800,7 +1809,7 @@ void RenderSSAO(CommandBuffer cmd, HDCamera hdCamera, ScriptableRenderContext re
                         var depthTexture = hdCamera.frameSettings.enableMSAA ? m_SharedRTManager.GetDepthValuesTexture() : m_SharedRTManager.GetDepthTexture();
 
                         HDUtils.CheckRTCreated(aoTarget.rt);
-                        postProcessLayer.BakeMSVOMap(cmd, camera, aoTarget, depthTexture, true, hdCamera.frameSettings.enableMSAA);
+                        postProcessLayer.BakeMSVOMap(cmd, camera, aoTarget, depthTexture, true);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: why enableMSAA disapear here?

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeEditor.Skin.cs [hunk 5]
```diff
@@ -1 +1,28 @@
...
+        internal static Color k_GizmoThemeColorExtentFace = new Color(255f / 255f, 229f / 255f, 148f / 255f, 45f / 255f);
+        internal static Color k_GizmoThemeColorInfluenceBlend = new Color(83f / 255f, 255f / 255f, 95f / 255f, 75f / 255f);
+        internal static Color k_GizmoThemeColorInfluenceBlendFace = new Color(83f / 255f, 255f / 255f, 95f / 255f, 17f / 255f);
+        internal static Color k_GizmoThemeColorInfluenceNormalBlend = new Color(0f / 255f, 229f / 255f, 255f / 255f, 80f / 255f);
+        internal static Color k_GizmoThemeColorInfluenceNormalBlendFace = new Color(0f / 255f, 229f / 255f, 255f / 255f, 36f / 255f);
+        internal static Color k_GizmoThemeColorProjection = new Color(0x00 / 255f, 0xE5 / 255f, 0xFF / 255f, 0x20 / 255f);
+        internal static Color k_GizmoThemeColorProjectionFace = new Color(0x00 / 255f, 0xE5 / 255f, 0xFF / 255f, 0x20 / 255f);
+        internal static Color k_GizmoThemeColorDisabled = new Color(0x99 / 255f, 0x89 / 255f, 0x59 / 255f, 0x10 / 255f);
+        internal static Color k_GizmoThemeColorDisabledFace = new Color(0x99 / 255f, 0x89 / 255f, 0x59 / 255f, 0x10 / 255f);
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeEditor.Skin.cs]: Does theses added static colors are used somewhere? I assume it is linked to Influence and should be in classe like InfluenceVolumeUI.Skin isn't it?

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/ReflectionProxyVolumeComponentUI.cs [hunk 6]
```diff
@@ -1,53 +1,25 @@
...
-        static ReflectionProxyVolumeComponentUI()
-        {
-            Inspector = CED.Select(
-                    (s, d, o) => s.proxyVolume,
-                    (s, d, o) => d.proxyVolume,
-                    ProxyVolumeUI.SectionShape
-                    );
-        }
-
+        #region Inspector
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/ReflectionProxyVolumeComponentUI.cs]: I think region is proscribed by c# reference 

And there is really no interest to do it in such a small class

File: com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.cs]: I see you rewrite the same as ExpandedState (same idea)
Note that ExpendedState  is used in CoreEditorDrawer (which have been moved in HDRP package). 
Could you merge both of theses class ?
[Can be done in another PR]

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbe.cs [hunk 8]
```diff
@@ -1,192 +1,194 @@
-using UnityEngine.Rendering;
-using UnityEngine.Serialization;
+using System;
 
 namespace UnityEngine.Experimental.Rendering.HDPipeline
 {
-    [ExecuteAlways]
-    public abstract class HDProbe : MonoBehaviour, ISerializationCallbackReceiver
+    [ExecuteInEditMode]
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbe.cs]: I think I read somthing about ExecuteInEditMode which is now obsolete and should be replaced by ExecuteAlways.
So why this change?

File: com.unity.render-pipelines.high-definition/Tests/Editor/PlanarReflectionProbe.MigrationTests.cs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+                    // No custom proxy here, so proxyToWorld = influenceToWorld
+                    var proxyToWorld = Matrix4x4.TRS(influencePositionWS, influenceRotationWS, Vector3.one);
+                    var mirrorPositionPS = (Vector3)proxyToWorld.inverse.MultiplyPoint(legacyProbeData.mirrorPositionWS);
+
+                    var instance = Object.Instantiate(prefab);
+
+                    var probe = instance.GetComponent<PlanarReflectionProbe>();
+                    prefab.SetActive(true);
+                    probe.enabled = true;
+
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Tests/Editor/PlanarReflectionProbe.MigrationTests.cs]: don't you need to wait a frame here in order to the probe.OnEnable be done ? (in order for the migration to be done if still done in OnEnable (unsure))

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeEditor.Skin.cs (hunk 5 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeEditor.Skin.cs]: Used skin in influence volume instead

File: com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.cs (hunk 7 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Core/EditorPrefBoolFlags.cs]: Will do in another PR

File: com.unity.render-pipelines.high-definition/Editor/Core/Gizmo/HierarchicalBox.cs [hunk 10]
```diff
@@ -67,27 +67,27 @@ namespace UnityEditor.Experimental.Rendering
     public class HierarchicalBox
     {
         const float k_HandleSizeCoef = 0.05f;
+        static Material k_Material_Cache;
+        static Material k_Material => (k_Material_Cache == null || k_Material_Cache.Equals(null) ? (k_Material_Cache = new Material(Shader.Find("Hidden/UnlitTransparentColored"))) : k_Material_Cache);
+        static Mesh k_MeshQuad_Cache;
+        static Mesh k_MeshQuad => k_MeshQuad_Cache == null || k_MeshQuad_Cache.Equals(null) ? (k_MeshQuad_Cache = Resources.GetBuiltinResource<Mesh>("Quad.fbx")) : k_MeshQuad_Cache;
```
[the remark this answers is not in the document]
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Core/Gizmo/HierarchicalBox.cs]: Those act exactly as const actually, the reason it is written like this is because we need  a lazy init to access Unityès mesh and material API.

So IMO, it is best to follow the const rule even though it is a lazy init

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDProbeUI.Skin.cs [hunk 11]
```diff
@@ -25,19 +20,14 @@ partial class HDProbeUI
...
-                    { ToolBar.NormalBlend,  EditorGUIUtility.IconContent("SceneViewOrtho", "|Modify the influence normal volume. (SHIFT+3)") },
-                    { ToolBar.CapturePosition,  EditorGUIUtility.IconContent("MoveTool", "|Change the Offset of the shape.") }
-                });
-            }
-        }
+            { ToolBar.InfluenceShape,  EditorGUIUtility.IconContent("EditCollider", "|Modify the base shape. (SHIFT+1)") },
+            { ToolBar.Blend,  EditorGUIUtility.IconContent("PreMatCube", "|Modify the influence volume. (SHIFT+2)") },
+            { ToolBar.NormalBlend,  EditorGUIUtility.IconContent("SceneViewOrtho", "|Modify the influence normal volume. (SHIFT+3)") },
+            { ToolBar.CapturePosition,  EditorGUIUtility.IconContent("MoveTool", "|Change the capture position.") },
+            { ToolBar.MirrorPosition,  EditorGUIUtility.IconContent("MoveTool", "|Change the mirror position.") },
```
[the remark this answers is not in the document]
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDProbeUI.Skin.cs]: Currently, CapturePosition and MirrorPosition are both exclusive (one for reflection probe, the other for Planar probes)

We can't merge both properties because the tooltips are differents

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeUI.cs [hunk 12]
```diff
@@ -1,21 +1,7 @@
...
-
 namespace UnityEditor.Experimental.Rendering.HDPipeline
 {
     internal partial class HDReflectionProbeUI : HDProbeUI
     {
-        internal HDReflectionProbeUI()
-        {
-            toolBars = new[] { ToolBar.InfluenceShape | ToolBar.Blend | ToolBar.NormalBlend, ToolBar.CapturePosition };
-        }
+
```
[the remark this answers is not in the document]
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDReflectionProbeUI.cs]: Will do in next pr

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/PlanarReflectionProbeUI.cs [hunk 13]
```diff
@@ -1,73 +1,7 @@
...
-                    )
-                );
-        }
-
-        internal PlanarReflectionProbeUI()
-        {
-            //remove normal edition tool and capture point for planar
-            toolBars = new[] { ToolBar.InfluenceShape | ToolBar.Blend, ToolBar.CapturePosition }; 
-        }
+
```
[the remark this answers is not in the document]
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/PlanarReflectionProbeUI.cs]: Will do in next pr

File: com.unity.render-pipelines.high-definition/Tests/Editor/PlanarReflectionProbe.MigrationTests.cs (hunk 9 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Tests/Editor/PlanarReflectionProbe.MigrationTests.cs]: No need to wait a frame because OnEnable is called right away

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbe.cs (hunk 8 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbe.cs]: Mismerge issue, I reverted the change

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/ReflectionProxyVolumeComponentUI.cs (hunk 6 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/ReflectionProxyVolumeComponentUI.cs]: pragma region were removed

File: com.unity.render-pipelines.high-definition/Runtime/Utilities/CameraSettings.cs [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+                cullingMask = -1,
+                useOcclusionCulling = true,
+                invertCulling = false
+            };
+
+            /// <summary>True when occlusion culling will be performed during rendering, false otherwise.</summary>
+            public bool useOcclusionCulling;
+            /// <summary>The mask for visible objects.</summary>
+            public int cullingMask;
+            /// <summary>True to invert face culling, false otherwise.</summary>
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/Utilities/CameraSettings.cs]: Not a big fan of this being bundled together with the two others. Mask and Occlusion are about culling of objects but invertCulling relates to face culling on the GPU which is not culling in the same sense. As is, it's a bit misleading.
I guess having a more explicit name would help a little bit (invertFaceCulling or something like that?)

**Comment by RemyUnity** [comment]
If "Auto Generate" is enabled in the Lighting window and that the folder that should contain the light data is missing in the project (folder placed aside the scene asset, and with the same name), then the editor is spamming the following error in the console : 
```
DirectoryNotFoundException: Destination directory not found: Assets\Scenes\samplescene
System.IO.File.Copy (System.String sourceFileName, System.String destFileName, System.Boolean overwrite) (at <ac823e2bb42b41bda67924a45a0173c3>:0)
UnityEngine.Experimental.Rendering.HDPipeline.HDBakedReflectionSystem.Tick (UnityEditor.Experimental.Rendering.SceneStateHash sceneStateHash, UnityEditor.Experimental.Rendering.IScriptableBakedReflectionSystemStageNotifier handle) (at L:/Gits/Graphics/com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDBakedReflectionSystem.cs:200)
UnityEditor.Experimental.Rendering.ScriptableBakedReflectionSystemWrapper.Internal_ScriptableBakedReflectionSystemWrapper_Tick (UnityEditor.Experimental.Rendering.SceneStateHash deps) (at C:/buildslave/unity/build/Editor/Mono/Camera/ScriptableBakedReflectionSystemWrapper.bindings.cs:84)
```

**Comment by RemyUnity** [comment]
Using the "Clear Baked Data" button from the Lighting window doesn't delete the baked cubemaps.

**Comment by RemyUnity** [comment]
When a new Reflection Probe is created, no settings of the influence volume can be changed, until the scene is saved and re-opened again.

**Comment by RemyUnity** [comment]
The reflection probe and proxy gizmos are not displayed like on master. Which one is correct ? : 
This branch :
image

Master:
image

**Comment by RSlysz** [comment]
Proxy (right) can be updated either like probe or like point light. It seams good.
However, gizmo of probe (non selected) should have same aspect than when editing them. It seams like a regression.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/InfluenceVolumeUI.Gizmos.cs [hunk 15]
```diff
@@ -19,55 +19,42 @@ public enum HandleType
...
-                        if ((showedHandle & HandleType.InfluenceNormal) != 0)
-                        {
-                            s.boxInfluenceNormalHandle.monoHandle = !s.data.editorAdvancedModeEnabled.boolValue;
-                            s.boxInfluenceNormalHandle.center = d.offset + d.boxBlendNormalOffset;
-                            s.boxInfluenceNormalHandle.size = d.boxSize + d.boxBlendNormalSize;
-                            s.boxInfluenceNormalHandle.DrawHull((editedHandle & HandleType.InfluenceNormal) != 0);
-                        }
-                        break;
+                    case InfluenceShape.Box: Gizmos.DrawWireCube(Vector3.zero, d.boxSize); break;
+                    case InfluenceShape.Sphere: Gizmos.DrawWireSphere(Vector3.zero, d.sphereRadius); break;
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/InfluenceVolumeUI.Gizmos.cs]: Regression on gizmo should be due to using Gizmos.DrawWireSphere instead of the DrawHull of HierarchicalSphere.

Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/Volume/InfluenceVolumeUI.Gizmos.cs]: fix pushed

**Comment by RemyUnity** [comment]
When a projection proxy with non null rotation is used, the capture position handle is not displayed where the preview is (or is in the inverse ?)

gizmo proxy gifski

**Comment by RemyUnity** [comment]
The "Height Preview" gizmos of the probe capture displays oddly, compared to the probe volume gizmo for example :
image

The two colors (red z-test off, green on) should probably be more close to avoid this effect.

**Comment by RemyUnity** [comment]
Maybe the Light Layer parameter should be grayed out when Light Layer Support is disabled in the HDRP Asset, like for the lights UI.
Reflection Probe :
image

Light:
image

File: com.unity.render-pipelines.high-definition/Runtime/Utilities/CameraSettings.cs (hunk 14 above)
Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/Runtime/Utilities/CameraSettings.cs]: Renamed from CameraSettings.Culling.invertCulling to CameraSettings.invertFaceCulling

**Comment by RemyUnity** [comment]
Reflection probes don't render with post processes, so the following setting should probably be removed/hidden from the Frame Settings Override :
image

**Comment by RemyUnity** [comment]
Using a custom rendering path for baking the reflection probe doesn't seems to work at all.
The following screenshot is the result of the baking after having disabled almost everything :
image

**Comment by RemyUnity** [comment]
The "Fullscreen Passthrough" option in the "Rendering Path" dropdown is useless in the context of baking probes:
image

Talked about it with @sebastienlagarde and @JulienIgnace-Unity and maybe the proper solution would be to replace this dropdown by two toogles to enable/disable the passthrough and the use of custom settings.

**Comment by RemyUnity** [comment]
The culling mask is displayed as an int value, instead of the classic layer mask dropdown:
image

**Comment by RemyUnity** [comment]
On rev https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/2392/commits/e3767430af9eac6de2f283e7c0631d4cb5698eca

The reflection probes gizmos don't show anymore outside of play mode.
And in play mode they are positioned at world origin.

**Comment by RemyUnity** [comment]
When the shape is set to sphere, no gizmo is shown until and editing mode (shape, blend distance, blend normal distance) is selected.
The shape gizmo should always be displayed when selected, like in box mode.
Maybe also the blend distance gizmo (lines only).

**Comment by RemyUnity** [comment]
The capture position "vertical helper" gizmo should take in account the gizmo size set in the scene view (that changes the small sphere preview)
probecaptureverticalgizmo

**Comment by RemyUnity** [comment]
# Tests:
✖ Type: Tested only baked, as realtime is not currently supported.
## Projection Settings
✔ Proxy Volume: Test baking with and without proxy volume assigned
✔ Use influence volume as projection
## Influence Volume
✔ Normal and Advance mode
✔ Shape: Box and Sphere
✔ Size
✔ Blend distance
✔ Blend normal distance
✔ Scene view manipulators
## Capture Settings
✔ Capture position
✔ Capture position scene manipulator
✔ Clear modes: sky / color / none
✔ Background Color
✔ Depth clear
✔ Volume layer mask
✔ Volume anchor override
✔ Use Occlusion Culling
✔ Culling Mask
✔ Clip Planes
✔ Rendering Path (Fullscreen passthrough should be remove tough)
## Custom Settings
✔ Light Layer
⁉  Multiplier: Value should not be allowed to be negative
⁉  Weight: Value should not be allowed to be negative

File: com.unity.render-pipelines.high-definition/CHANGELOG.md [hunk 16]
```diff
@@ -162,6 +163,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/CHANGELOG.md]: be careful, it is not expected that you change otherr line of theog

Comment (fredpointzero) [reply com.unity.render-pipelines.high-definition/CHANGELOG.md]: It was a merge error. I removed those duplicate lines

File: com.unity.render-pipelines.high-definition/Runtime/Core/CoreResources/GPUCopy.compute [hunk 17]
```diff
@@ -1,6 +1,6 @@
 // Autogenerated file. Do not edit by hand
 
-#pragma only_renderers d3d11 ps4 xboxone vulkan metal switch
+#pragma only_renderers d3d11 ps4 xboxone vulkan metal
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Core/CoreResources/GPUCopy.compute]: why removed switch ? Please add it back in another PR

**Comment by sebastienlagarde** [comment]
The PR have several visual effect graph editor test failing 
https://artifacts-ui.bf.unity3d.com/proj57-Test_Editor_-_Windows/27639004_01_12_2018_12_06_38_%2B0000/logs/TestResults_editor_VisualEffectGraph.xml

you will need to fix them to be able to merge (not we are on lastest master)

**Comment by sebastienlagarde** [comment]
Katana is green, current failing test are from master and unrelated to the PR. Merging.
